plugins/modules/dcnm_service_policy.py

```diff
@@ -0,0 +1,1962 @@
+        state.
+
+        Parameters:
+            cfg (dict): The config from playbook
+
+        Returns:
+            None
+        """
+
+        sp_delete_spec = dict(
+            node_name=dict(required=True, type="str"),
+            rp_name=dict(required=False, type="str", default=""),
+            name=dict(required=False, type="str", default=""),
+        )
+
+        sp_info, invalid_params = validate_list_of_dicts(cfg, sp_delete_spec)
+        if invalid_params:
+            if cfg[0].get("name", " ") != " ":
+                mesg = "Invalid parameters in playbook: {}".format(
+                    "while processing Service Policy -  "
+                    + cfg[0]["name"]
+                    + ", "
+                    + "".join(invalid_params)
+                )
+            else:
+                mesg = "Invalid parameters in playbook: {}".format(
+                    "while processing Service Policy -  Unknown, "
+                    + "".join(invalid_params)
+                )
+            self.module.fail_json(msg=mesg)
+
+        self.sp_info.extend(sp_info)
+
+    def dcnm_sp_validate_query_state_input(self, cfg):
+
+        """
+        Playbook input will be different for differnt states. This routine validates the query state
+        input. This routine updates self.sp_info with validated playbook information related to query
+        state.
+
+        Parameters:
+            cfg (dict): The config from playbook
+
+        Returns:
+           None
+        """
+
+        sp_query_spec = dict(
+            node_name=dict(required=True, type="str"),
+            name=dict(required=False, type="str", default="None"),
+        )
+
+        sp_info, invalid_params = validate_list_of_dicts(cfg, sp_query_spec)
+        if invalid_params:
+            mesg = "Invalid parameters in playbook: {}".format(
+                "while processing Service Policy -  "
+                + cfg[0]["name"]
+                + ", "
+                + "\n".join(invalid_params)
+            )
+            self.module.fail_json(msg=mesg)
+
+        self.sp_info.extend(sp_info)
+
+    def dcnm_sp_get_service_node_type(self, sp):
+
+        path = (
+            "/appcenter/Cisco/elasticservice/elasticservice-api/fabrics/"
+            + self.module.params["attached_fabric"]
+            + "/service-nodes/"
+            + sp["node_name"]
+        )
+
+        retries = 0
+        while retries <= 30:
+
+            retries += 1
+            resp = dcnm_send(self.module, "GET", path)
+
+            if resp and (resp["RETURN_CODE"] == 200) and resp["DATA"]:
+                return resp["DATA"].get("type", None)
+            else:
+                self.dcnm_sp_check_unauthorized_error_in_resp(resp)
+                time.sleep(10)
+                continue
+        self.changed_dict[0]["debugs"].append({"GET_SNODE_TYPE": resp})
+        return "Firewall"
```

Comment:
This return confuses me.  Why are we returning a hardcoded value here?

Reply:
Should not. It should be from the response. Removed it.